### ISSUE TITLE
support webmock 3.0

### ISSUE THREAD

**Issue by pravi** [abonas/kubeclient#260]
We'd like to use webmock 3.0 in debian, but some tests are failing.

```
$ bundle install
Fetching gem metadata from https://rubygems.org/............
Fetching version metadata from https://rubygems.org/..
Fetching dependency metadata from https://rubygems.org/..
Resolving dependencies...
Using rake 10.5.0
Using public_suffix 2.0.5
Using unf_ext 0.0.7.4
Using http-form_data 1.0.3
Using http_parser.rb 0.6.0
Using recursive-open-struct 1.0.0
Using mime-types-data 3.2016.0521
Using netrc 0.11.0
Using minitest 5.10.3
Using safe_yaml 1.0.4
Using hashdiff 0.3.5
Using vcr 3.0.3
Using ast 2.3.0
Using powerpack 0.1.1
Using ruby-progressbar 1.8.1
Using bundler 1.15.1
Using rainbow 2.2.2
Using addressable 2.5.1
Using unf 0.1.4
Using mime-types 3.1
Using crack 0.4.3
Using parser 2.4.0.0
Using domain_name 0.5.20170404
Fetching webmock 3.0.1 (was 1.24.6)
Installing webmock 3.0.1 (was 1.24.6)
Using astrolabe 1.3.1
Using http-cookie 1.0.3
Using rubocop 0.30.0
Using http 0.9.8
Using rest-client 2.0.2
Using kubeclient 2.3.0 from source at `.`
Bundle complete! 7 Gemfile dependencies, 30 gems now installed.
Bundled gems are installed into ./vendor/bundle.
The latest bundler is 1.15.3, but you are currently running 1.15.1.
To update, run `gem install bundler`
pravi@nishumbha:~/forge/debian/git/pkg-ruby-extras/ruby-kubeclient$ bundle exec rake
Run options: --seed 37337

# Running:

..........................E....F.E.......................DEPRECATED: Use assert_nil if expecting nil from /home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/test/test_service.rb:42. This will fail in Minitest 6.
.........................

Finished in 0.525047s, 156.1766 runs/s, 676.1302 assertions/s.

  1) Error:
KubeClientTest#test_api_basic_auth_back_comp_success:
WebMock::NetConnectNotAllowedError: Real HTTP connections are disabled. Unregistered request: GET http://localhost:8080/api/v1 with headers {'Accept'=>'*/*', 'Accept-Encoding'=>'gzip, deflate', 'Authorization'=>'Basic dXNlcm5hbWU6cGFzc3dvcmQ=', 'Host'=>'localhost:8080', 'User-Agent'=>'rest-client/2.0.2 (linux-gnu x86_64) ruby/2.3.3p222'}

You can stub this request with the following snippet:

stub_request(:get, "http://localhost:8080/api/v1").
  with(headers: {'Accept'=>'*/*', 'Accept-Encoding'=>'gzip, deflate', 'Authorization'=>'Basic dXNlcm5hbWU6cGFzc3dvcmQ=', 'Host'=>'localhost:8080', 'User-Agent'=>'rest-client/2.0.2 (linux-gnu x86_64) ruby/2.3.3p222'}).
  to_return(status: 200, body: "", headers: {})

registered request stubs:

stub_request(:get, "http://username:password@localhost:8080/api/v1/pods")
stub_request(:get, "http://username:password@localhost:8080/api/v1")

============================================================
    /home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/vendor/bundle/gems/webmock-3.0.1/lib/webmock/http_lib_adapters/net_http.rb:114:in `request'
    /home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/vendor/bundle/gems/rest-client-2.0.2/lib/restclient/request.rb:448:in `net_http_do_request'
    /home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/vendor/bundle/gems/rest-client-2.0.2/lib/restclient/request.rb:721:in `block in transmit'
    /home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/vendor/bundle/gems/webmock-3.0.1/lib/webmock/http_lib_adapters/net_http.rb:123:in `start_without_connect'
    /home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/vendor/bundle/gems/webmock-3.0.1/lib/webmock/http_lib_adapters/net_http.rb:150:in `start'
    /home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/vendor/bundle/gems/rest-client-2.0.2/lib/restclient/request.rb:715:in `transmit'
    /home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/vendor/bundle/gems/rest-client-2.0.2/lib/restclient/request.rb:145:in `execute'
    /home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/vendor/bundle/gems/rest-client-2.0.2/lib/restclient/request.rb:52:in `execute'
    /home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/vendor/bundle/gems/rest-client-2.0.2/lib/restclient/resource.rb:51:in `get'
    /home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/lib/kubeclient/common.rb:437:in `block in fetch_entities'
    /home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/lib/kubeclient/common.rb:99:in `handle_exception'
    /home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/lib/kubeclient/common.rb:437:in `fetch_entities'
    /home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/lib/kubeclient/common.rb:427:in `load_entities'
    /home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/lib/kubeclient/common.rb:111:in `discover'
    /home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/lib/kubeclient/common.rb:78:in `method_missing'
    /home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/test/test_kubeclient.rb:446:in `test_api_basic_auth_back_comp_success'

  2) Failure:
KubeClientTest#test_api_basic_auth_failure [/home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/test/test_kubeclient.rb:468]:
[KubeException] exception expected, not
Class: <WebMock::NetConnectNotAllowedError>
Message: <"Real HTTP connections are disabled. Unregistered request: GET http://localhost:8080/api/v1 with headers {'Accept'=>'*/*', 'Accept-Encoding'=>'gzip, deflate', 'Authorization'=>'Basic dXNlcm5hbWU6cGFzc3dvcmQ=', 'Host'=>'localhost:8080', 'User-Agent'=>'rest-client/2.0.2 (linux-gnu x86_64) ruby/2.3.3p222'}\n\nYou can stub this request with the following snippet:\n\nstub_request(:get, \"http://localhost:8080/api/v1\").\n  with(headers: {'Accept'=>'*/*', 'Accept-Encoding'=>'gzip, deflate', 'Authorization'=>'Basic dXNlcm5hbWU6cGFzc3dvcmQ=', 'Host'=>'localhost:8080', 'User-Agent'=>'rest-client/2.0.2 (linux-gnu x86_64) ruby/2.3.3p222'}).\n  to_return(status: 200, body: \"\", headers: {})\n\nregistered request stubs:\n\nstub_request(:get, \"http://username:password@localhost:8080/api/v1\")\n\n============================================================">
---Backtrace---
/home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/vendor/bundle/gems/webmock-3.0.1/lib/webmock/http_lib_adapters/net_http.rb:114:in `request'
/home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/vendor/bundle/gems/rest-client-2.0.2/lib/restclient/request.rb:448:in `net_http_do_request'
/home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/vendor/bundle/gems/rest-client-2.0.2/lib/restclient/request.rb:721:in `block in transmit'
/home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/vendor/bundle/gems/webmock-3.0.1/lib/webmock/http_lib_adapters/net_http.rb:123:in `start_without_connect'
/home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/vendor/bundle/gems/webmock-3.0.1/lib/webmock/http_lib_adapters/net_http.rb:150:in `start'
/home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/vendor/bundle/gems/rest-client-2.0.2/lib/restclient/request.rb:715:in `transmit'
/home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/vendor/bundle/gems/rest-client-2.0.2/lib/restclient/request.rb:145:in `execute'
/home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/vendor/bundle/gems/rest-client-2.0.2/lib/restclient/request.rb:52:in `execute'
/home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/vendor/bundle/gems/rest-client-2.0.2/lib/restclient/resource.rb:51:in `get'
/home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/lib/kubeclient/common.rb:437:in `block in fetch_entities'
/home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/lib/kubeclient/common.rb:99:in `handle_exception'
/home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/lib/kubeclient/common.rb:437:in `fetch_entities'
/home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/lib/kubeclient/common.rb:427:in `load_entities'
/home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/lib/kubeclient/common.rb:111:in `discover'
/home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/lib/kubeclient/common.rb:78:in `method_missing'
/home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/test/test_kubeclient.rb:468:in `block in test_api_basic_auth_failure'
---------------

  3) Error:
KubeClientTest#test_api_basic_auth_success:
WebMock::NetConnectNotAllowedError: Real HTTP connections are disabled. Unregistered request: GET http://localhost:8080/api/v1 with headers {'Accept'=>'*/*', 'Accept-Encoding'=>'gzip, deflate', 'Authorization'=>'Basic dXNlcm5hbWU6cGFzc3dvcmQ=', 'Host'=>'localhost:8080', 'User-Agent'=>'rest-client/2.0.2 (linux-gnu x86_64) ruby/2.3.3p222'}

You can stub this request with the following snippet:

stub_request(:get, "http://localhost:8080/api/v1").
  with(headers: {'Accept'=>'*/*', 'Accept-Encoding'=>'gzip, deflate', 'Authorization'=>'Basic dXNlcm5hbWU6cGFzc3dvcmQ=', 'Host'=>'localhost:8080', 'User-Agent'=>'rest-client/2.0.2 (linux-gnu x86_64) ruby/2.3.3p222'}).
  to_return(status: 200, body: "", headers: {})

registered request stubs:

stub_request(:get, "http://username:password@localhost:8080/api/v1/pods")
stub_request(:get, "http://username:password@localhost:8080/api/v1")

============================================================
    /home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/vendor/bundle/gems/webmock-3.0.1/lib/webmock/http_lib_adapters/net_http.rb:114:in `request'
    /home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/vendor/bundle/gems/rest-client-2.0.2/lib/restclient/request.rb:448:in `net_http_do_request'
    /home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/vendor/bundle/gems/rest-client-2.0.2/lib/restclient/request.rb:721:in `block in transmit'
    /home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/vendor/bundle/gems/webmock-3.0.1/lib/webmock/http_lib_adapters/net_http.rb:123:in `start_without_connect'
    /home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/vendor/bundle/gems/webmock-3.0.1/lib/webmock/http_lib_adapters/net_http.rb:150:in `start'
    /home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/vendor/bundle/gems/rest-client-2.0.2/lib/restclient/request.rb:715:in `transmit'
    /home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/vendor/bundle/gems/rest-client-2.0.2/lib/restclient/request.rb:145:in `execute'
    /home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/vendor/bundle/gems/rest-client-2.0.2/lib/restclient/request.rb:52:in `execute'
    /home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/vendor/bundle/gems/rest-client-2.0.2/lib/restclient/resource.rb:51:in `get'
    /home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/lib/kubeclient/common.rb:437:in `block in fetch_entities'
    /home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/lib/kubeclient/common.rb:99:in `handle_exception'
    /home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/lib/kubeclient/common.rb:437:in `fetch_entities'
    /home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/lib/kubeclient/common.rb:427:in `load_entities'
    /home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/lib/kubeclient/common.rb:111:in `discover'
    /home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/lib/kubeclient/common.rb:78:in `method_missing'
    /home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/test/test_kubeclient.rb:423:in `test_api_basic_auth_success'

82 runs, 355 assertions, 1 failures, 2 errors, 0 skips
rake aborted!
Command failed with status (1): [ruby -I"lib:test" -I"/home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/vendor/bundle/gems/rake-10.5.0/lib" "/home/pravi/forge/debian/git/pkg-ruby-extras/ruby-kubeclient/vendor/bundle/gems/rake-10.5.0/lib/rake/rake_test_loader.rb" "test/test*.rb" ]

Tasks: TOP => default => test
(See full trace by running task with --trace)
```

**Comment by cben** [comment]
since #242, kubeclient depends on webmock ~> 3.0.1.
This has landed in kubeclient 3.0.0 but not in 2.5.2.  This implies 2.x branch isn't tested on ruby 2.4, 2.5. It's definitely possible to backport this to a 2.x release, if desired.
Closing, but please reopen (or open new issues) if there is anything we can help for packaging :heart: 

I see you're maintainer of https://packages.debian.org/sid/ruby-kubeclient — let me know if you like to be cc'd on kubeclient decisions about dependency bumps & releases.

**Comment by pravi** [comment]
@cben thanks! I'm happy to see your interest in debian packaging (which is a rare thing these days of language specific package managers being the exclusive channel for many developers). 

The only package that depends on ruby-kubeclient is gitlab which still requires ~> 2.2.0. [I have asked them to update](https://gitlab.com/gitlab-org/gitlab-ce/issues/43911) and when they do we'll update the package to 3.0. I don't think a backport is necessary at this point (we may need it only if gitlab cannot move to 3.0 for some reason).

It would be very helpful if you follow SemVer for your releases and if you do that already, mention it in README. https://wiki.debian.org/Teams/Ruby/UpstreamPledge documents response from different projects. There has been some negative response, but overall it is still positive.

 I'd be interested in any breaking changes due to releases or dependency bumps (major version bumps), so cc me.
